geode-core/src/main/java/org/apache/geode/cache/client/internal/pooling/ConnectionManagerImpl.java

```diff
@@ -277,146 +294,77 @@ public Connection borrowConnection(long acquireTimeout)
   @Override
   public Connection borrowConnection(ServerLocation server, long acquireTimeout,
       boolean onlyUseExistingCnx) throws AllConnectionsInUseException, NoAvailableServersException {
-    lock.lock();
-    try {
-      if (shuttingDown) {
-        throw new PoolCancelledException();
-      }
-      for (Iterator<PooledConnection> itr = availableConnections.iterator(); itr.hasNext();) {
-        PooledConnection nextConnection = itr.next();
-        try {
-          nextConnection.activate();
-          if (nextConnection.getServer().equals(server)) {
-            itr.remove();
-            return nextConnection;
-          }
-          nextConnection.passivate(false);
-        } catch (ConnectionDestroyedException ex) {
-          // someone else already destroyed this connection so ignore it
-          // but remove it from availableConnections
-        }
-        // Fix for 41516. Before we let this method exceed the max connections
-        // by creating a new connection, we need to make sure that they're
-        // aren't bogus connections sitting in the available connection list
-        // otherwise, the length of that list might exceed max connections,
-        // but with some bad connections. That can cause members to
-        // get a bad connection but have no permits to create a new connection.
-        if (nextConnection.shouldDestroy()) {
-          itr.remove();
-        }
-      }
-
-      if (onlyUseExistingCnx) {
-        throw new AllConnectionsInUseException();
-      }
 
-      // We need to create a connection. Reserve space for it.
-      connectionCount++;
-    } finally {
-      lock.unlock();
+    if (shuttingDown.get()) {
+      throw new PoolCancelledException();
     }
 
-    PooledConnection connection = null;
-    try {
-      Connection plainConnection = connectionFactory.createClientToServerConnection(server, false);
-      connection = addConnection(plainConnection);
-    } catch (GemFireSecurityException e) {
-      throw new ServerOperationException(e);
-    } finally {
-      // if we failed, release the space we reserved for our connection
+    for (int i = availableConnections.size(); i > 0; i--) {
```

Comment:
Now that availableConnections is a concurrent queue and other threads can be concurrently adding and removing from it, it seems wrong to have the size of the queue at one point in time limit our iterations. The other complication is that we add items we took off the head back to the tail in this loop so we could actually see them more than once. We should rethink this code now that it no longer holds a lock during the iteration.

Reply:
Yeah, I think I have a solution for that using the iterator. 

We expected in this solution that the size was just an arbitrary limit on the times through the pool, though one very close to the entire pool. We were trying to work around a race potentially with trying to pull one off near the front of the queue while it is also being polled by as the queue drains.